src/main/java/com/scalyr/integrations/kafka/Event.java

```diff
@@ -169,18 +187,48 @@ public int estimatedSerializedBytes() {
       return estimatedSizeBytes;
     }
 
-    int size = getMessage() == null ? 0 : getMessage().length();
+    int size = Strings.isNullOrEmpty(getMessage()) ? 0 : estimateEscapedStringSize(getMessage());
     size += getTopic().length();
     size += EVENT_SERIALIZATION_OVERHEAD_BYTES;
 
     if (getAdditionalAttrs() != null) {
       size += getAdditionalAttrs().entrySet().stream()
-        .mapToInt(entry -> entry.getKey().length() + (entry.getValue() == null ? 0 : entry.getValue().toString().length())).sum();
+        .mapToInt(entry -> entry.getKey().length() + (entry.getValue() == null ? 0 : estimateEscapedStringSize(entry.getValue().toString()))).sum();
     }
     estimatedSizeBytes = size;
     return estimatedSizeBytes;
   }
 
+  /**
+   * Estimate the escaped String length for JSON strings.
+   * JSON payload requires extra escaping which needs to be taken into the size estimate.
+   *
+   * If it is not JSON, then use the String length as the estimate.
+   * This trades off some inaccuracies where the String may contain quotable characters for faster performance.
+   * @return Estimated escaped String size
+   */
+  private int estimateEscapedStringSize(String s) {
+    if (s == null) {
+      return 0;
+    }
+    if (s.length() < 2) {
+      return s.length();
+    }
+
+    if (s.charAt(0) == '{' && s.charAt(s.length() - 1) == '}') {
+      return s.length() + countJsonEscapedCharacters(s);
+    }
+
+    return s.length();
+  }
+
+  private int countJsonEscapedCharacters(String s) {
```

Comment:
Do we have any idea on how much overhead this adds for JSON objects of various sizes?

If you have some time, I think it would be good to add a test case which times this function for a large JSON object, just so we have some idea on the overhead / timing involved.

Reply:
I tested this with a ~15K JSON payload from a customer.  I ran `countJsonEscapedCharacters` 100_000 times.  It got an average of 0.19 milliseconds per ~15K JSON.  ~0.2 milliseconds seems like reasonable overhead.